### ISSUE TITLE
Invalid maximum heap size: -Xmx4096M

### ISSUE THREAD

**Issue by paulas** [ruboto/ruboto#749]
Hi! 

I have heap error
Is there 32bit OS problem?

Can you give some advice?

__________________________________________
[user@lacky-thing testapp3]$ rake
Picked up _JAVA_OPTIONS: -Xms256m -Xmx512m
ant debug
Picked up _JAVA_OPTIONS: -Xms256m -Xmx512m
Buildfile: /home/user/testapp3/build.xml
Trying to override old definition of task dex-helper

-set-mode-check:

-set-debug-files:

-check-env:
 [checkenv] Android SDK Tools Revision 24.3.3
 [checkenv] Installed at /home/user/android-sdk-linux

-setup:
     [echo] Project Name: Testapp3
  [gettype] Project Type: Application

-set-debug-mode:

-debug-obfuscation-check:

-pre-build:

-build-setup:
[getbuildtools] Using latest Build Tools: 22.0.1
     [echo] Resolving Build Target for Testapp3...
[gettarget] Project Target:   Android 4.2.2
[gettarget] API level:        17
     [echo] ----------
     [echo] Creating output directories if needed...
    [mkdir] Created dir: /home/user/testapp3/bin/rsObj
    [mkdir] Created dir: /home/user/testapp3/bin/rsLibs
     [echo] ----------
     [echo] Resolving Dependencies for Testapp3...
[dependency] Library dependencies:
[dependency] No Libraries
[dependency] 
[dependency] ------------------
     [echo] ----------
     [echo] Building Libraries with 'debug'...
   [subant] No sub-builds to iterate on

-code-gen:
[mergemanifest] No changes in the AndroidManifest files.
     [echo] Handling aidl files...
     [aidl] No AIDL files to compile.
     [echo] ----------
     [echo] Handling RenderScript files...
     [echo] ----------
     [echo] Handling Resources...
     [aapt] No changed resources. R.java and Manifest.java untouched.
     [echo] ----------
     [echo] Handling BuildConfig class...
[buildconfig] Generating BuildConfig class.

-pre-compile:

-compile:
    [javac] Compiling 1 source file to /home/user/testapp3/bin/classes
    [javac] warning: [options] source value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] target value 1.5 is obsolete and will be removed in a future release
    [javac] warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
    [javac] 3 warnings

-post-compile:

-obfuscate:

-dex:
    [apply] Invalid maximum heap size: -Xmx4096M
    [apply] The specified size exceeds the maximum representable size.
    [apply] Error: Could not create the Java Virtual Machine.
    [apply] Error: A fatal exception has occurred. Program will exit.

BUILD FAILED
/home/user/android-sdk-linux/tools/ant/build.xml:888: The following error occurred while executing this line:
/home/user/android-sdk-linux/tools/ant/build.xml:890: The following error occurred while executing this line:
/home/user/android-sdk-linux/tools/ant/build.xml:902: The following error occurred while executing this line:
/home/user/testapp3/build.xml:174: apply returned: 1

Total time: 4 seconds
rake aborted!
Command failed with status (1): [ant debug...]
/home/user/testapp3/rakelib/ruboto.rake:966:in `build_apk'
/home/user/testapp3/rakelib/ruboto.rake:552:in `block in <top (required)>'
Tasks: TOP => default => debug => /home/user/testapp3/bin/Testapp3-debug.apk
(See full trace by running task with --trace)
__________________________________________

**Comment by vanbumi** [comment]
anyone can fix this problem?  I have same case

**Comment by donv** [comment]
Hi @paulas and @vanbumi !

Thanks for the report.  This looks like a 32 bit issue.

http://stackoverflow.com/questions/909018/avoiding-initial-memory-heap-size-error

Try setting the ´ DX_HEAP_SIZE` environment value to a valid size like 2048:

```bash
export DX_HEAP_SIZE=2048
```

Our logic for this is here:

https://github.com/ruboto/ruboto/blob/master/assets/rakelib/ruboto.rake#L558

A pull request that adds logic for 32 bit systems would be welcome :)

Alternatively this could be a setting in `ruboto.yml`:

```yaml
###################################
#
#dx_ heap_size
#   
#    Set the heap size used by the "dx" command in megabytes.
#    Must be large enough to process your app, but lower than your available memory.
#    On 32-bit systems  value between 2048 and 3072 should work.
#    Default is 4096
#
# dx_heap_size: 4096

````

https://github.com/ruboto/ruboto/blob/master/assets/ruboto.yml

Feel free to fork and fix.  Pull requests very welcome :)

**Comment by petervojtek** [comment]
I have this issue as well (32bit ubuntu, zsh).
playing with DX_HEAP_SIZE or exporting Xmx into shell was not working for me.

what finally helped was to open following file `~/android-sdk-linux/build-tools/23.0.0/dx`

and rewrite following line:
```
defaultMx="-Xmx4096M"
```
to some lower value.

**Comment by donv** [comment]
Hi @petervojtek !

Did you find the Ruboto code that sets the `defaultMx="-XmxnnnnM"` setting in the `dx`script?  Maybe we could add something that avoids setting it too high?  Currently we check to see if we need to **increase** the setting.  Maybe we should check if we should **lower** it for 32-bit systems as well.  Feel free to submit a pull request if you see how to improve it.

**Comment by donv** [comment]
I think this has been addressed in #779 .

**Comment by eladkarako** [comment]
If you are using 32bit installation of java, it is best to use a smaller value, 
for example `-Xmx2048m`.
if you are using a 64bit installation of java you can use this value, 
and if you have any issues, you can explicitly use `-d32` argument for java (not to be placed in `JAVA_TOOL_OPTIONS`).

on Windows batch file `run.cmd` you can use:

```cmd
@echo off
chcp 65001 2>nul >nul
set "JAVA_TOOL_OPTIONS="
set "JAVA_TOOL_OPTIONS=%JAVA_TOOL_OPTIONS% -Xmn512m -Xms512m -Xmx2048m"
set "JAVA_TOOL_OPTIONS=%JAVA_TOOL_OPTIONS% -XX:+UseParallelGC -XX:ParallelGCThreads=2 -XX:+UseG1GC -Xverify:none"
set "JAVA_TOOL_OPTIONS=%JAVA_TOOL_OPTIONS% -Dfile.encoding=UTF8 -Duser.language=en"

set "ARCHITECTURE=-d32"
if ["%PROCESSOR_ARCHITECTURE%"] EQU ["AMD64"] (
  set "ARCHITECTURE=-d64"
)

::the JAVA_TOOL_OPTIONS is an environment variable JVM is looking for, and there is no need of putting it as an argument each run...
call java.exe %ARCHITECTURE% -jar foo.jar
pause
exit /b %ErrorLevel%
```
